diff --git a/GLideN64/src/GraphicsDrawer.h b/GLideN64/src/GraphicsDrawer.h
index 66c8f62..b1b20b0 100644
--- a/GLideN64/src/GraphicsDrawer.h
+++ b/GLideN64/src/GraphicsDrawer.h
@@ -17,8 +17,9 @@ namespace graphics {
 struct CachedTexture;
 struct FrameBuffer;

-#define VERTBUFF_SIZE 256U
-#define ELEMBUFF_SIZE 1024U
+constexpr u32 VERTBUFF_SIZE = 256U;
+constexpr u32 VERTBUFF_MASK = VERTBUFF_SIZE - 1;
+constexpr u32 ELEMBUFF_SIZE = 1024U;

 constexpr f32 SCREEN_SIZE_DIM = 640.0f;
 constexpr u32 MIPMAP_TILE_WIDTH = 256u;
@@ -136,7 +137,7 @@ public:

 	bool isRejected(u32 _v0, u32 _v1, u32 _v2) const;

-	SPVertex & getVertex(u32 _v) { return triangles.vertices[_v]; }
+	SPVertex & getVertex(u32 _v) { return triangles.vertices[_v&VERTBUFF_MASK]; }

 	SPVertex * getVertexPtr(u32 _v) { return triangles.vertices.data() + _v; }

diff --git a/GLideN64/src/gSP.cpp b/GLideN64/src/gSP.cpp
index a1b2c3d..e4f5a6b 100644
--- a/GLideN64/src/gSP.cpp
+++ b/GLideN64/src/gSP.cpp
@@ -1563,6 +1563,11 @@ bool gSPCullVertices( u32 v0, u32 vn )
 	if (vn < v0) {
 		// Aidyn Chronicles - The First Mage seems to pass parameters in reverse order.
 		std::swap(v0, vn);
 	}
+
+	if (vn >= VERTBUFF_SIZE)
+		// Wrong input. DL is wrong, let's cull it.
+		return true;
+
 	u32 clip = 0;
 	GraphicsDrawer & drawer = dwnd().getDrawer();
 	for (u32 i = v0; i <= vn; ++i) {
